### PR TITLE
Utility methods update

### PR DESCRIPTION
**Pull request by CptCrispyCrunchy**
-Added a method to center a template around its geometrical center in the XZ plan.

-Moved the `SpawnBlockRegionComponent.getBoundingBox()` to a new utility class to prepare for the engine update

### CODE REVIEW

File: src/main/java/org/terasology/structureTemplates/util/BlockRegionToolbox.java [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+import org.terasology.structureTemplates.util.transform.BlockRegionMovement;
+
+import java.util.List;
+
+
+/**
+ * A collection of little helper to ease the work with structure templates.
+ **/
+
+public class BlockRegionToolbox {
```
Comment (skaldarnar) [inline src/main/java/org/terasology/structureTemplates/util/BlockRegionToolbox.java]: I would name it `BlockRegionUtils`

File: src/test/java/org/terasology/structureTemplates/components/BlockRegionBoundingBoxTest.java [hunk 2]
```diff
@@ -24,6 +24,7 @@
 import org.terasology.world.block.BlockUri;
 
 import static org.junit.Assert.assertEquals;
+import static org.terasology.structureTemplates.util.BlockRegionToolbox.getBoundingBox;
```
Comment (skaldarnar) [inline src/test/java/org/terasology/structureTemplates/components/BlockRegionBoundingBoxTest.java]: Try to avoid static imports.

File: src/main/java/org/terasology/structureTemplates/util/BlockRegionToolbox.java [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+
+    public static BlockRegionMovement setOnCenterXZ(SpawnBlockRegionsComponent spawnBlockRegionsComponent) {
```
Comment (skaldarnar) [inline src/main/java/org/terasology/structureTemplates/util/BlockRegionToolbox.java]: Probably add a general method to center along any (combination) of axis and provide convenience methods like this one.

File: src/test/java/org/terasology/structureTemplates/components/BlockRegionBoundingBoxTest.java (hunk 2 above)
Comment (CptCrispyCrunchy) [inline src/test/java/org/terasology/structureTemplates/components/BlockRegionBoundingBoxTest.java]: fixed :)

File: src/main/java/org/terasology/structureTemplates/util/BlockRegionToolbox.java (hunk 1 above)
Comment (CptCrispyCrunchy) [inline src/main/java/org/terasology/structureTemplates/util/BlockRegionToolbox.java]: done :)

**Comment by CptCrispyCrunchy** [comment]
@flo : ping! DynamicCities would need that PR to run stable on the event tomorrow. Would be cool if you can take a look at it :)

File: src/main/java/org/terasology/structureTemplates/util/BlockRegionUtilities.java [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+import java.util.List;
+
+
+/**
+ * A collection of little helper to ease the work with structure templates.
+ **/
+
+public class BlockRegionUtilities {
+
+    public static BlockRegionMovement setOnCenterXZ(SpawnBlockRegionsComponent spawnBlockRegionsComponent) {
```
Comment (flo) [inline src/main/java/org/terasology/structureTemplates/util/BlockRegionUtilities.java]: I think it should not be named "set" as it is not a setter and does not set something.

Also there might be the need to determine the center for other reasons than the creation of BlockRegionMovement. So I would go for a method say like "Vector3i determineBottomCenterOf".

File: src/test/java/org/terasology/structureTemplates/components/BlockRegionBoundingBoxTest.java [hunk 5]
```diff
@@ -46,9 +47,9 @@ public void testBoundingBox() {
         blockRegionA.regionsToFill.add(regionA);
         blockRegionA.regionsToFill.add(regionB);
         blockRegionA.regionsToFill.add(regionC);
-        Region3i bb = blockRegionA.getBoundingBox();
-        assertEquals(bb.max(), new Vector3i(6, 1, 5));
-        assertEquals(bb.min(), new Vector3i(2, 0, -1));
+        Region3i bb = BlockRegionUtilities.getBoundingBox(blockRegionA);
+        Assert.assertEquals(bb.max(), new Vector3i(6, 1, 5));
```
Comment (flo) [inline src/test/java/org/terasology/structureTemplates/components/BlockRegionBoundingBoxTest.java]: Was this change intended. I prefer usually assertEquals without the Assert infront.

File: src/main/java/org/terasology/structureTemplates/util/BlockRegionUtilities.java [hunk 6]
```diff
@@ -0,0 +1,72 @@
...
+import org.terasology.structureTemplates.components.SpawnBlockRegionsComponent;
+import org.terasology.structureTemplates.util.transform.BlockRegionMovement;
+
+import java.util.List;
+
+
+/**
+ * A collection of little helper to ease the work with structure templates.
+ **/
+
```
Comment (flo) [inline src/main/java/org/terasology/structureTemplates/util/BlockRegionUtilities.java]: Is the empty line here intended?

It is however good that you added a javadoc. However I would suggest that you be more specific as the whole  module is about structure templates.

File: src/main/java/org/terasology/structureTemplates/components/SpawnBlockRegionsComponent.java [hunk 7]
```diff
@@ -31,35 +31,6 @@
 public class SpawnBlockRegionsComponent implements Component {
     public List<RegionToFill> regionsToFill = new ArrayList<>();
 
-    public Region3i getBoundingBox() {
```
Comment (flo) [inline src/main/java/org/terasology/structureTemplates/components/SpawnBlockRegionsComponent.java]: thanks for moving it to a utility  class, that is much better

File: src/main/java/org/terasology/structureTemplates/util/BlockRegionUtilities.java (hunk 4 above)
Comment (skaldarnar) [inline src/main/java/org/terasology/structureTemplates/util/BlockRegionUtilities.java]: I still would go for an even more general version, like
```Java
public static Vector3i determineCenterPointAlongAxisOf(SpawnBlockRegionsComponent component, Axis... axis)
```
What do you think? `determineBottomCenterPointOf` could be implemented as:
```Java
public static Vector3i determineBottomCenterPointOf(SpawnBlockRegionsComponent component) {
    return determineCenterPointAlongAxisFor(component, Axis.X, Axis.Z);
}
```

P.S.: I don't think we have general axis descriptors yet, only [`Direction`](https://github.com/MovingBlocks/Terasology/blob/develop/engine/src/main/java/org/terasology/math/Direction.java).

File: src/test/java/org/terasology/structureTemplates/components/BlockRegionBoundingBoxTest.java (hunk 5 above)
Comment (skaldarnar) [inline src/test/java/org/terasology/structureTemplates/components/BlockRegionBoundingBoxTest.java]: oh, that accounts probably accounts to me advising to avoid static imports... I guess static imports are fine for `Assert` in unit tests, but `BlockRegionUtils` should not be statically imported imho.

Comment (flo) [inline src/test/java/org/terasology/structureTemplates/components/BlockRegionBoundingBoxTest.java]: yes, especially getBoudningBox should not be statically imported as it could be confused with a local getter.

File: src/main/java/org/terasology/structureTemplates/util/BlockRegionUtilities.java (hunk 4 above)
Comment (flo) [inline src/main/java/org/terasology/structureTemplates/util/BlockRegionUtilities.java]: I can't think of a scenario where you would need the general version of it. I would say there is more important stuff than a general version of it, since there would be also more to test.

Comment (skaldarnar) [inline src/main/java/org/terasology/structureTemplates/util/BlockRegionUtilities.java]: Hm, maybe my thoughts were too complicated. I had some scenarios in mind were you would like to align a structure to something else, maybe placing it under the ceiling of a big cave, or aligning with neighbouring structures. But that can also be achieved with the simple version for bottomCenter and the general bounding box I guess...

File: src/main/java/org/terasology/structureTemplates/util/BlockRegionUtilities.java (hunk 6 above)
Comment (CptCrispyCrunchy) [inline src/main/java/org/terasology/structureTemplates/util/BlockRegionUtilities.java]: Oh the empty line was by accident. 
As for being more specific: I guess I adding some method descriptions instead of that line could be more helpful then.

File: src/main/java/org/terasology/structureTemplates/util/BlockRegionUtilities.java (hunk 4 above)
Comment (CptCrispyCrunchy) [inline src/main/java/org/terasology/structureTemplates/util/BlockRegionUtilities.java]: As for the naming: It sort of "sets" all blocks to be centered around 0,0. But you are right, it is probably a bit confusing like this.

Adding additional centering methods could be a thing for the future but as long as it doesn't needed I deem it unnecessary too. Though I guess it could be used for some cool clipping stuff later on. I would implement just additional methods though (would only be 2 more) instead of defining an axis class etc.

**Comment by CptCrispyCrunchy** [comment]
@flo latest commit should have everything right now. Refactored method name, added comments to the methods instead of the class, even removed empty lines :D and `determineBottomCenter()` returns a Vector3i now.

**Comment by flo** [comment]
merged
